File: 744. Find Smallest Letter Greater Than Target/main.c
#include <stdbool.h>
#include <stdio.h>

char nextGreatestLetter(char *letters, int lettersSize, char target) {
    if (target < letters[0] || target >= letters[lettersSize - 1]) {
        return letters[0];
    }

    int left = 0;
    int right = lettersSize - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2;

        if (letters[mid] == target) {
            if (letters[mid + 1] != target) {
                return letters[mid + 1];
            }
            left = mid + 1;
        } else if (letters[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return letters[left];
}

char nextGreatestLetter(char *letters, int lettersSize, char target) {
    int left = 0;
    int right = lettersSize - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2;

        if (letters[mid] <= target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return letters[left % lettersSize];
}

File: 744. Find Smallest Letter Greater Than Target/main.rs
struct Solution;

impl Solution {
    pub fn next_greatest_letter(letters: Vec<char>, target: char) -> char {
        if letters.first().gt(&Some(&target)) || letters.last().le(&Some(&target)) {
            return letters[0];
        }

        let mut left = 0;
        let mut right = letters.len() as i32 - 1;

        while left <= right {
            let mid = left + (right - left) / 2;

            if letters[mid as usize] == target {
                if letters[mid as usize + 1] > target {
                    return letters[mid as usize + 1];
                }

                left = mid + 1;
            } else if letters[mid as usize] < target {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        letters[left as usize]
    }
}

impl Solution {
    pub fn next_greatest_letter(letters: Vec<char>, target: char) -> char {
        let mut left = 0;
        let mut right = letters.len() as i32 - 1;

        while left <= right {
            let mid = left + (right - left) / 2;

            if letters[mid as usize] <= target {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        letters[left as usize % letters.len()]
    }
}
